### ISSUE TITLE
str.cat fails on multiple columns

### ISSUE THREAD

**Issue by sephib** [dask/dask#5621]
Hi,  
We encountered the following issue:
(`dask` version 2.8 and `pandas` 0.25.3)  

```
import dask.dataframe as dd
import dask.datasets as ds
import pandas as pd

df = ds.timeseries().reset_index()
```

When using `str.cat` on a single column there is no problem

`df['timestamp'].astype(str).str.cat(df['name'].astype(str), sep='_').head()`

When running with multiple columns we get a KeyError:  
`df['timestamp'].astype(str).str.cat(df[['id', 'name']].astype(str), sep='_').head()`   

see below:  

<details>
--------------------------------------------------------------------------

KeyError                                  Traceback (most recent call last)

/usr/local/lib/python3.6/dist-packages/pandas/core/indexes/base.py in get_loc(self, key, method, tolerance)
   2896             try:
-> 2897                 return self._engine.get_loc(key)
   2898             except KeyError:

pandas/_libs/index.pyx in pandas._libs.index.IndexEngine.get_loc()

pandas/_libs/index.pyx in pandas._libs.index.IndexEngine.get_loc()

pandas/_libs/hashtable_class_helper.pxi in pandas._libs.hashtable.PyObjectHashTable.get_item()

pandas/_libs/hashtable_class_helper.pxi in pandas._libs.hashtable.PyObjectHashTable.get_item()

KeyError: 0

During handling of the above exception, another exception occurred:

KeyError                                  Traceback (most recent call last)

5 frames

/usr/local/lib/python3.6/dist-packages/pandas/core/indexes/base.py in get_loc(self, key, method, tolerance)
   2897                 return self._engine.get_loc(key)
   2898             except KeyError:
-> 2899                 return self._engine.get_loc(self._maybe_cast_indexer(key))
   2900         indexer = self.get_indexer([key], method=method, tolerance=tolerance)
   2901         if indexer.ndim > 1 or indexer.size > 1:

pandas/_libs/index.pyx in pandas._libs.index.IndexEngine.get_loc()

pandas/_libs/index.pyx in pandas._libs.index.IndexEngine.get_loc()

pandas/_libs/hashtable_class_helper.pxi in pandas._libs.hashtable.PyObjectHashTable.get_item()

pandas/_libs/hashtable_class_helper.pxi in pandas._libs.hashtable.PyObjectHashTable.get_item()

KeyError: 0
</details>   
When converting the code to `Pandas` there is no error:    

```
dfp = df.compute()
dfp['timestamp'].astype(str).str.cat(dfp[['id', 'name']].astype(str), sep='_').head()
```

**Comment by TomAugspurger** [comment]
Yes, it seems like we just haven't implemented this case yet. A PR doing so would be welcome.

I suspect we'll want to replace the `valid_types` with checks to `is_series_like`, `is_index_like` or `is_dataframe_like`.

**Comment by sephib** [comment]
Hi,
We are suggesting for the PR to check for the general iterable requirement, and keep the the `valid types`.  
How about:   

```
from collections import Iterable

@derived_from(pd.core.strings.StringMethods)
def cat(self, others=None, sep=None, na_rep=None):
    from .core import Series, Index

    if others is None:
        raise NotImplementedError("x.str.cat() with `others == None`")

    valid_types = (Series, Index, pd.Series, pd.Index)
    if isinstance(others, valid_types):
        others = [others]
    elif isinstance(others, Iterable) and (not all(isinstance(a, valid_types) for a in others)):
        raise TypeError("others must be of Series/Index (or an Iterable of Series/Index)")

    return self._series.map_partitions(
        str_cat, *others, sep=sep, na_rep=na_rep, meta=self._series._meta
    )
```

If the `import collection` is problematic we can check only for list/tuple

**Comment by sephib** [comment]
Hi,
Following your comment - bellow is our suggested PR  

```
@derived_from(pd.core.strings.StringMethods)
def cat(self, others=None, sep=None, na_rep=None):
    from .core import DataFrame, Series, Index       

    if others is None:
        raise NotImplementedError("x.str.cat() with `others == None`")

    valid_types = (Series, Index, pd.Series, pd.Index)
    is_dataframe = (DataFrame, pd.DataFrame)
    
    if isinstance(others, valid_types):
        others = [others]
    elif isinstance(others, is_dataframe):
        raise TypeError("others must be Series/Index and not dataframe")
    elif not all(isinstance(a, valid_types) for a in others):
        raise TypeError("others must be Series/Index")

    return self._series.map_partitions(
        str_cat, *others, sep=sep, na_rep=na_rep, meta=self._series._meta
    )
```

And update the test with: 
```
for o in ["foo", ["foo"], df, ddf]:        
    with pytest.raises(TypeError):
        ddf.str_col.str.cat(o)
```  

If this is the direction we will submit the PR

**Comment by TomAugspurger** [comment]
Probably easiest to discuses in a PR at this point.

I would recommend using `is_series_like`, etc, rather than a list of types we consider valid.

**Comment by TomAugspurger** [comment]
The KeyError is coming from `iter(dask.dataframe.DataFrame)` being incorrect. I'll put up a fix for that.

**Comment by TomAugspurger** [comment]
With #5719, I get the "expected" behavior of raising a TypeError.

```
In [2]:  df['timestamp'].astype(str).str.cat(df[['id', 'name']].astype(str), sep='_').head()
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
<ipython-input-2-0e58ce323dbf> in <module>
----> 1 df['timestamp'].astype(str).str.cat(df[['id', 'name']].astype(str), sep='_').head()

~/sandbox/dask/dask/dataframe/accessor.py in cat(self, others, sep, na_rep)
    145             others = [others]
    146         elif not all(isinstance(a, valid_types) for a in others):
--> 147             raise TypeError("others must be Series/Index")
    148
    149         return self._series.map_partitions(

TypeError: others must be Series/Index
```

So that at least is clearer than the KeyError. We can still implement proper support for DataFrames though.

**Comment by jsignell** [comment]
Closing this issue since the failure is clear now.
